From 2cb6ea5dc7599c49cee406a92afd9ddf4a59b879 Mon Sep 17 00:00:00 2001
From: Jianxiong Pan <jianxiong.pan@amlogic.com>
Date: Thu, 28 Nov 2024 15:15:07 +0800
Subject: [PATCH] mm: add amlogic vmap stack for arm64. [1/1]

PD#SWPL-188077

Problem:
optimizate kernel stack.

Solution:
enable amlogic vmap stack.

Verify:
local.

Change-Id: I295e3164d240a22eadb4e5b1854edcad12c7833e
Signed-off-by: Jianxiong Pan <jianxiong.pan@amlogic.com>
---
 Makefile                                   |   7 ++
 arch/arm64/Kconfig                         |   2 +-
 arch/arm64/include/asm/hw_breakpoint.h     |   6 +
 arch/arm64/include/asm/memory.h            |   3 +
 arch/arm64/include/asm/stacktrace.h        |   4 +
 arch/arm64/include/asm/stacktrace/common.h |  12 ++
 arch/arm64/kernel/Makefile                 |   4 +
 arch/arm64/kernel/entry-common.c           |  16 +++
 arch/arm64/kernel/entry.S                  | 124 +++++++++++++++++++++
 arch/arm64/kernel/hibernate.c              |   6 +
 arch/arm64/kernel/hw_breakpoint.c          |   7 ++
 arch/arm64/kernel/irq.c                    |   5 +
 arch/arm64/kernel/smp.c                    |  10 ++
 arch/arm64/kernel/stacktrace.c             |  55 +++++++++
 arch/arm64/kernel/suspend.c                |   7 ++
 arch/arm64/mm/kasan_init.c                 |   4 +
 fs/proc/meminfo.c                          |   9 ++
 include/linux/sched.h                      |   6 +
 include/linux/sched/task_stack.h           |   8 ++
 include/linux/vmalloc.h                    |   6 +
 kernel/fork.c                              |  25 +++++
 lib/Kconfig.debug                          |   1 +
 mm/kmemleak.c                              |  25 +++++
 23 files changed, 351 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8237627e2d3e..f58101c6092c 100644
--- a/Makefile
+++ b/Makefile
@@ -888,9 +888,11 @@ ifdef CONFIG_READABLE_ASM
 KBUILD_CFLAGS += -fno-reorder-blocks -fno-ipa-cp-clone -fno-partial-inlining
 endif
 
+ifndef CONFIG_AMLOGIC_STACKPROTECTOR
 stackp-flags-y                                    := -fno-stack-protector
 stackp-flags-$(CONFIG_STACKPROTECTOR)             := -fstack-protector
 stackp-flags-$(CONFIG_STACKPROTECTOR_STRONG)      := -fstack-protector-strong
+endif
 
 KBUILD_CFLAGS += $(stackp-flags-y)
 
@@ -900,6 +902,11 @@ KBUILD_RUSTFLAGS += $(KBUILD_RUSTFLAGS-y)
 ifdef CONFIG_FRAME_POINTER
 KBUILD_CFLAGS	+= -fno-omit-frame-pointer -fno-optimize-sibling-calls
 KBUILD_RUSTFLAGS += -Cforce-frame-pointers=y
+ifdef CONFIG_AMLOGIC_VMAP
+ifdef CONFIG_ARM64
+KBUILD_CFLAGS += -mno-omit-leaf-frame-pointer
+endif
+endif
 else
 # Some targets (ARM with Thumb2, for example), can't be built with frame
 # pointers.  For those, we don't have FUNCTION_TRACER automatically
diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 7d4566807b43..44629e6be503 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -189,7 +189,7 @@ config ARM64
 	select HAVE_ARCH_THREAD_STRUCT_WHITELIST
 	select HAVE_ARCH_TRACEHOOK
 	select HAVE_ARCH_TRANSPARENT_HUGEPAGE
-	select HAVE_ARCH_VMAP_STACK
+	select HAVE_ARCH_VMAP_STACK if !AMLOGIC_VMAP
 	select HAVE_ARM_SMCCC
 	select HAVE_ASM_MODVERSIONS
 	select HAVE_EBPF_JIT
diff --git a/arch/arm64/include/asm/hw_breakpoint.h b/arch/arm64/include/asm/hw_breakpoint.h
index bd81cf17744a..f23c11d1ad83 100644
--- a/arch/arm64/include/asm/hw_breakpoint.h
+++ b/arch/arm64/include/asm/hw_breakpoint.h
@@ -148,9 +148,15 @@ static inline int get_num_brps(void)
 static inline int get_num_wrps(void)
 {
 	u64 dfr0 = read_sanitised_ftr_reg(SYS_ID_AA64DFR0_EL1);
+#ifdef CONFIG_AMLOGIC_VMAP
+	return 1 +
+		cpuid_feature_extract_unsigned_field(dfr0,
+						ID_AA64DFR0_EL1_WRPs_SHIFT) - 2;
+#else
 	return 1 +
 		cpuid_feature_extract_unsigned_field(dfr0,
 						ID_AA64DFR0_EL1_WRPs_SHIFT);
+#endif
 }
 
 #ifdef CONFIG_CPU_PM
diff --git a/arch/arm64/include/asm/memory.h b/arch/arm64/include/asm/memory.h
index 7eef2496fba9..0c3a09c85ebd 100644
--- a/arch/arm64/include/asm/memory.h
+++ b/arch/arm64/include/asm/memory.h
@@ -129,6 +129,9 @@
 #endif
 
 #define THREAD_SIZE		(UL(1) << THREAD_SHIFT)
+#ifdef CONFIG_AMLOGIC_VMAP
+#define THREAD_START_SP         (THREAD_SIZE - 16)
+#endif
 
 /*
  * By aligning VMAP'd stacks to 2 * THREAD_SIZE, we can detect overflow by
diff --git a/arch/arm64/include/asm/stacktrace.h b/arch/arm64/include/asm/stacktrace.h
index 66ec8caa6ac0..c100255ec93c 100644
--- a/arch/arm64/include/asm/stacktrace.h
+++ b/arch/arm64/include/asm/stacktrace.h
@@ -20,6 +20,10 @@
 extern void dump_backtrace(struct pt_regs *regs, struct task_struct *tsk,
 			   const char *loglvl);
 
+#ifdef CONFIG_AMLOGIC_VMAP
+//extern __always_inline int kunwind_next(struct kunwind_state *state);
+#endif
+
 DECLARE_PER_CPU(unsigned long *, irq_stack_ptr);
 
 static inline struct stack_info stackinfo_get_irq(void)
diff --git a/arch/arm64/include/asm/stacktrace/common.h b/arch/arm64/include/asm/stacktrace/common.h
index f63dc654e545..696056f7b713 100644
--- a/arch/arm64/include/asm/stacktrace/common.h
+++ b/arch/arm64/include/asm/stacktrace/common.h
@@ -11,6 +11,10 @@
 
 #include <linux/types.h>
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
+
 struct stack_info {
 	unsigned long low;
 	unsigned long high;
@@ -94,6 +98,14 @@ static inline int unwind_consume_stack(struct unwind_state *state,
 	if (stackinfo_on_stack(&state->stack, sp, size))
 		goto found;
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	/*
+	 * keep search stack for task
+	 */
+	if (on_vmap_stack(sp, &state->stack))
+		goto found;
+#endif
+
 	next = unwind_find_next_stack(state, sp, size);
 	if (!next)
 		return -EINVAL;
diff --git a/arch/arm64/kernel/Makefile b/arch/arm64/kernel/Makefile
index 2b112f3b7510..411d0b295240 100644
--- a/arch/arm64/kernel/Makefile
+++ b/arch/arm64/kernel/Makefile
@@ -13,6 +13,10 @@ CFLAGS_REMOVE_return_address.o = $(CC_FLAGS_FTRACE)
 # checks due to randomize_kstack_offset.
 CFLAGS_REMOVE_syscall.o	 = -fstack-protector -fstack-protector-strong
 CFLAGS_syscall.o	+= -fno-stack-protector
+ifeq ($(CONFIG_AMLOGIC_PTR_AUTH_KERNEL),y)
+CFLAGS_REMOVE_entry.o = -mbranch-protection=pac-ret
+CFLAGS_REMOVE_entry-common.o = -mbranch-protection=pac-ret
+endif
 
 # When KASAN is enabled, a stack trace is recorded for every alloc/free, which
 # can significantly impact performance. Avoid instrumenting the stack trace
diff --git a/arch/arm64/kernel/entry-common.c b/arch/arm64/kernel/entry-common.c
index 4d541320e011..353f2f6c120b 100644
--- a/arch/arm64/kernel/entry-common.c
+++ b/arch/arm64/kernel/entry-common.c
@@ -32,6 +32,10 @@
 #include <trace/hooks/gic.h>
 #include <trace/hooks/dtask.h>
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
 
 /*
  * Handle IRQ/context state management when entering from kernel mode.
  * Before this function is called it is not safe to call regular kernel code,
@@ -526,6 +530,9 @@ asmlinkage void noinstr el1h_64_sync_handler(struct pt_regs *regs)
 	switch (ESR_ELx_EC(esr)) {
 	case ESR_ELx_EC_DABT_CUR:
 	case ESR_ELx_EC_IABT_CUR:
+	#ifdef CONFIG_AMLOGIC_VMAP
+		regs = handle_vmap_fault(read_sysreg(far_el1), esr, regs);
+	#endif
 		el1_abort(regs, esr);
 		break;
 	/*
@@ -537,6 +544,9 @@ asmlinkage void noinstr el1h_64_sync_handler(struct pt_regs *regs)
 		break;
 	case ESR_ELx_EC_SYS64:
 	case ESR_ELx_EC_UNKNOWN:
+	#ifdef CONFIG_AMLOGIC_VMAP
+		regs = switch_vmap_context(regs);
+	#endif
 		el1_undef(regs, esr);
 		break;
 	case ESR_ELx_EC_BTI:
@@ -546,6 +556,9 @@ asmlinkage void noinstr el1h_64_sync_handler(struct pt_regs *regs)
 	case ESR_ELx_EC_SOFTSTP_CUR:
 	case ESR_ELx_EC_WATCHPT_CUR:
 	case ESR_ELx_EC_BRK64:
+	#ifdef CONFIG_AMLOGIC_VMAP
+		regs = switch_vmap_context(regs);
+	#endif
 		el1_dbg(regs, esr);
 		break;
 	case ESR_ELx_EC_FPAC:
@@ -590,6 +603,9 @@ static void noinstr el1_interrupt(struct pt_regs *regs,
 
 asmlinkage void noinstr el1h_64_irq_handler(struct pt_regs *regs)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	regs = switch_vmap_context(regs);
+#endif
 	el1_interrupt(regs, handle_arch_irq);
 }
 
diff --git a/arch/arm64/kernel/entry.S b/arch/arm64/kernel/entry.S
index 189ce50055d1..3a80145c1482 100644
--- a/arch/arm64/kernel/entry.S
+++ b/arch/arm64/kernel/entry.S
@@ -332,7 +332,11 @@ alternative_else_nop_endif
 	*/
 	.endm
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	.macro  kernel_exit, el, swap = 0
+#else
 	.macro	kernel_exit, el
+#endif /* CONFIG_AMLOGIC_VMAP */
 	.if	\el != 0
 	disable_daif
 	.endif
@@ -445,6 +449,20 @@ alternative_else_nop_endif
 	.endif
 
 	ldr	lr, [sp, #S_LR]
+#ifdef CONFIG_AMLOGIC_VMAP
+	.if	\el != 0
+	/* restore context sp and per-cpu vmap stack */
+	.if     \swap == 1
+	stp     x19, x20, [sp]
+	mov     x20, sp
+	add     x19, x20, #PT_REGS_SIZE
+	msr     DBGWVR3_EL1, x19
+	mrs     x19, DBGWVR2_EL1
+	mov     sp,  x19
+	ldp     x19, x20, [x20]
+	.endif
+	.endif
+#endif /* CONFIG_AMLOGIC_VMAP */
 	add	sp, sp, #PT_REGS_SIZE		// restore sp
 
 	.if \el == 0
@@ -569,8 +587,34 @@ SYM_CODE_END(__bad_stack)
 #endif /* CONFIG_VMAP_STACK */
 
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	.macro entry_handler el:req, ht:req, regsize:req, label:req, vmap = 0
+#else
 	.macro entry_handler el:req, ht:req, regsize:req, label:req
+#endif
 SYM_CODE_START_LOCAL(el\el\ht\()_\regsize\()_\label)
+#ifdef CONFIG_AMLOGIC_VMAP
+	.if \vmap == 1
+	/*
+	 * register using:
+	 * DBGWVR2_EL1: temp register and back up for sp_el1 of exception
+	 * DBGWVR3_EL1: always point to per-cpu vmap stack
+	 * switch sp_el1 to per-cpu vmap stack and using DBGWVR2_EL1
+	 * to back up sp_el1 under exception
+	 */
+	msr     DBGWVR2_EL1, x29
+	mrs     x29, DBGWVR3_EL1
+	sub     x29, x29, #PT_REGS_SIZE
+	msr     DBGWVR3_EL1, x29
+	stp     x19, x20, [x29]
+	mov     x19, sp
+	mrs     x20, DBGWVR2_EL1
+	msr     DBGWVR2_EL1, x19
+	mov     sp,  x29
+	mov     x29, x20
+	ldp     x19, x20, [sp]
+	.endif
+#endif /* CONFIG_AMLOGIC_VMAP */
 	kernel_entry \el, \regsize
 	mov	x0, sp
 	bl	el\el\ht\()_\regsize\()_\label\()_handler
@@ -590,8 +634,14 @@ SYM_CODE_END(el\el\ht\()_\regsize\()_\label)
 	entry_handler	1, t, 64, fiq
 	entry_handler	1, t, 64, error
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	/* pad vmap option */
+	entry_handler   1, h, 64, sync, 1
+	entry_handler   1, h, 64, irq,  1
+#else
 	entry_handler	1, h, 64, sync
 	entry_handler	1, h, 64, irq
+#endif
 	entry_handler	1, h, 64, fiq
 	entry_handler	1, h, 64, error
 
@@ -1102,3 +1152,77 @@ SYM_CODE_START(__sdei_handler_abort)
 SYM_CODE_END(__sdei_handler_abort)
 NOKPROBE(__sdei_handler_abort)
 #endif /* CONFIG_ARM_SDE_INTERFACE */
+
+#ifdef CONFIG_AMLOGIC_VMAP
+SYM_CODE_START(__setup_vmap_stack)
+	ldr		x10, =vmap_stack
+	add		x10, x10, x0
+	mov		x0, x10
+	mov		x1, #0
+	mov		x2, #THREAD_SIZE
+	mov		x17, lr
+#ifdef CONFIG_KASAN
+	bl		__memset		/* clear stack buffer */
+#else
+	bl		memset			/* clear stack buffer */
+#endif
+	mov		lr, x17
+	mov		x0, #THREAD_START_SP
+	add		x10, x10, x0		/* set stack top */
+	msr		DBGWVR3_EL1, x10
+	ret
+SYM_CODE_END(__setup_vmap_stack)
+
+SYM_CODE_START(__vmap_exit)
+	mov		sp, x0
+	kernel_exit 1, 1	/* return from vmap stack fault */
+SYM_CODE_END(__vmap_exit)
+
+	/* switch conext from vmap handler to normal handler */
+SYM_CODE_START(switch_vmap_context)
+	/*
+	 * Not a vmap fault, copy context saved in per-cpu vmap stack
+	 * to task stack, then switch stack back to task stack under
+	 * exception
+	 *          -------
+	 *          |     |  vmap context
+	 *          |     |
+	 *   x0 ->  |-----|
+	 *          |     |  function stack (diff is x3)
+	 *   sp ->  -------
+	 */
+	mov	x1, sp
+	sub	x3, x0, x1
+	mrs	x0, DBGWVR2_EL1			// load raw context sp
+	sub	x0, x0, x3
+	add	x2, x3, #PT_REGS_SIZE
+	stp	x19, x20, [sp, #-16]!
+	stp	x30, xzr, [sp, #-16]!
+	mov	x19, x2
+	mov	x20, x3
+#ifdef CONFIG_KASAN
+	bl	__memcpy
+#else
+	bl	memcpy
+#endif
+	add	x17, x0, x19
+	add	x19, x20, #S_SP
+	str	x17, [x0, x19]			// restore sp in context
+	add	x29, x0, x20			// old fp
+
+	mrs	x16, DBGWVR3_EL1
+	add	x16, x16, #PT_REGS_SIZE
+	msr	DBGWVR3_EL1, x16		// restore pcp vmap stack
+
+	mov	x17, x0				// sp in function context
+	add	x0,  x0, x20			// real pt-regs
+
+	ldp	x30, xzr, [sp], #16
+	ldp	x19, x20, [sp], #16
+	mov	sp, x17
+	add	x29, x29, #S_STACKFRAME
+	str	x29, [sp]			// restore frame pointer
+	mov	x29, sp
+	ret
+SYM_CODE_END(switch_vmap_context)
+#endif /* CONFIG_AMLOGIC_VMAP */
diff --git a/arch/arm64/kernel/hibernate.c b/arch/arm64/kernel/hibernate.c
index 0b43875e76f5..6a14b23fa180 100644
--- a/arch/arm64/kernel/hibernate.c
+++ b/arch/arm64/kernel/hibernate.c
@@ -34,6 +34,9 @@
 #include <asm/trans_pgd.h>
 #include <asm/virt.h>
 #include <trace/hooks/bl_hib.h>
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
 
 /*
  * Hibernate core relies on this value being 0 on resume, and marks it
@@ -356,6 +359,9 @@ int swsusp_arch_suspend(void)
 		sleep_cpu = smp_processor_id();
 		ret = swsusp_save();
 	} else {
+	#ifdef CONFIG_AMLOGIC_VMAP
+		__setup_vmap_stack(my_cpu_offset);
+	#endif
 		/* Clean kernel core startup/idle code to PoC*/
 		dcache_clean_inval_poc((unsigned long)__mmuoff_data_start,
 				    (unsigned long)__mmuoff_data_end);
diff --git a/arch/arm64/kernel/hw_breakpoint.c b/arch/arm64/kernel/hw_breakpoint.c
index 722ac45f9f7b..8d3768906afd 100644
--- a/arch/arm64/kernel/hw_breakpoint.c
+++ b/arch/arm64/kernel/hw_breakpoint.c
@@ -122,6 +122,13 @@ NOKPROBE_SYMBOL(read_wb_reg);
 
 static void write_wb_reg(int reg, int n, u64 val)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	/* avoid write DBGWVR since we use it for special purpose */
+	if ((reg + n) >= (AARCH64_DBG_REG_WVR + 2) &&
+		(reg + n) < AARCH64_DBG_REG_WCR) {
+		return;
+	}
+#endif
 	switch (reg + n) {
 	GEN_WRITE_WB_REG_CASES(AARCH64_DBG_REG_BVR, AARCH64_DBG_REG_NAME_BVR, val);
 	GEN_WRITE_WB_REG_CASES(AARCH64_DBG_REG_BCR, AARCH64_DBG_REG_NAME_BCR, val);
diff --git a/arch/arm64/kernel/irq.c b/arch/arm64/kernel/irq.c
index 85087e2df564..149296788e78 100644
--- a/arch/arm64/kernel/irq.c
+++ b/arch/arm64/kernel/irq.c
@@ -64,7 +64,12 @@ static void __init init_irq_stacks(void)
 }
 #else
 /* irq stack only needs to be 16 byte aligned - not IRQ_STACK_SIZE aligned. */
+#ifdef CONFIG_AMLOGIC_VMAP
+DEFINE_PER_CPU_ALIGNED(unsigned long [IRQ_STACK_SIZE/sizeof(long)], irq_stack)
+	__aligned(PAGE_SIZE);
+#else
 DEFINE_PER_CPU_ALIGNED(unsigned long [IRQ_STACK_SIZE/sizeof(long)], irq_stack);
+#endif
 
 static void init_irq_stacks(void)
 {
diff --git a/arch/arm64/kernel/smp.c b/arch/arm64/kernel/smp.c
index 1f9755ed5532..15ad0056aa90 100644
--- a/arch/arm64/kernel/smp.c
+++ b/arch/arm64/kernel/smp.c
@@ -57,6 +57,10 @@
 #undef CREATE_TRACE_POINTS
 #include <trace/hooks/debug.h>
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
+
 EXPORT_TRACEPOINT_SYMBOL_GPL(ipi_raise);
 EXPORT_TRACEPOINT_SYMBOL_GPL(ipi_entry);
 EXPORT_TRACEPOINT_SYMBOL_GPL(ipi_exit);
@@ -213,6 +217,9 @@ asmlinkage notrace void secondary_start_kernel(void)
 	const struct cpu_operations *ops;
 	unsigned int cpu = smp_processor_id();
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	__setup_vmap_stack(my_cpu_offset);
+#endif
 	/*
 	 * All kernel threads share the same mm context; grab a
 	 * reference and switch to it.
@@ -467,6 +474,9 @@ void __init smp_prepare_boot_cpu(void)
 	 */
 	set_my_cpu_offset(per_cpu_offset(smp_processor_id()));
 
+#ifdef CONFIG_AMLOGIC_VMAP
+	__setup_vmap_stack(my_cpu_offset);
+#endif
 	cpuinfo_store_boot_cpu();
 	setup_boot_cpu_features();
 
diff --git a/arch/arm64/kernel/stacktrace.c b/arch/arm64/kernel/stacktrace.c
index 9c444b15e6d7..e14cf957b090 100644
--- a/arch/arm64/kernel/stacktrace.c
+++ b/arch/arm64/kernel/stacktrace.c
@@ -20,6 +20,10 @@
 #include <asm/stack_pointer.h>
 #include <asm/stacktrace.h>
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
+
 /*
  * Kernel unwind state
  *
@@ -297,12 +301,59 @@ noinline noinstr void arch_bpf_stack_walk(bool (*consume_entry)(void *cookie, u6
 }
 EXPORT_SYMBOL_GPL(arch_stack_walk);
 
+#ifdef CONFIG_AMLOGIC_VMAP
+static noinline notrace void aml_dump_backtrace_entry(void *cookie,
+		struct task_struct *task, struct pt_regs *regs)
+{
+	struct stack_info stacks[] = {
+		stackinfo_get_task(task),
+		STACKINFO_CPU(irq),
+#if defined(CONFIG_VMAP_STACK)
+		STACKINFO_CPU(overflow),
+#endif
+#if defined(CONFIG_VMAP_STACK) && defined(CONFIG_ARM_SDE_INTERFACE)
+		STACKINFO_SDEI(normal),
+		STACKINFO_SDEI(critical),
+#endif
+#ifdef CONFIG_EFI
+		STACKINFO_EFI,
+#endif
+	};
+	struct kunwind_state state = {
+		.common = {
+			.stacks = stacks,
+			.nr_stacks = ARRAY_SIZE(stacks),
+		},
+	};
+
+	if (regs) {
+		if (task != current)
+			return;
+		kunwind_init_from_regs(&state, regs);
+	} else if (task == current) {
+		kunwind_init_from_caller(&state);
+	} else {
+		kunwind_init_from_task(&state, task);
+	}
+	
+	while(1) {
+		int ret;
+
+		dump_backtrace_entry_vmap(state.common.pc, state.common.fp,
+				(unsigned long)task->stack, (char *)cookie);
+		ret = kunwind_next(&state);
+		if (ret < 0)
+			break;
+	}
+}
+#else
 static bool dump_backtrace_entry(void *arg, unsigned long where)
 {
 	char *loglvl = arg;
 	printk("%s %pSb\n", loglvl, (void *)where);
 	return true;
 }
+#endif
 
 void dump_backtrace(struct pt_regs *regs, struct task_struct *tsk,
 		    const char *loglvl)
@@ -319,7 +370,11 @@ void dump_backtrace(struct pt_regs *regs, struct task_struct *tsk,
 		return;
 
 	printk("%sCall trace:\n", loglvl);
+#ifdef CONFIG_AMLOGIC_VMAP
+	aml_dump_backtrace_entry((void *)loglvl, tsk, regs);
+#else
 	arch_stack_walk(dump_backtrace_entry, (void *)loglvl, tsk, regs);
+#endif
 
 	put_task_stack(tsk);
 }
diff --git a/arch/arm64/kernel/suspend.c b/arch/arm64/kernel/suspend.c
index eaaff94329cd..14e36d0f0cf1 100644
--- a/arch/arm64/kernel/suspend.c
+++ b/arch/arm64/kernel/suspend.c
@@ -18,6 +18,9 @@
 #include <asm/mmu_context.h>
 #include <asm/smp_plat.h>
 #include <asm/suspend.h>
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
 
 /*
  * This is allocated by cpu_suspend_init(), and used to store a pointer to
@@ -137,6 +140,7 @@ int cpu_suspend(unsigned long arg, int (*fn)(unsigned long))
 	 */
 	arm_cpuidle_save_irq_context(&context);
 
+
 	ct_cpuidle_enter();
 
 	if (__cpu_suspend_enter(&state)) {
@@ -155,6 +159,9 @@ int cpu_suspend(unsigned long arg, int (*fn)(unsigned long))
 
 		ct_cpuidle_exit();
 	} else {
+	#ifdef CONFIG_AMLOGIC_VMAP
+		__setup_vmap_stack(my_cpu_offset);
+	#endif
 		ct_cpuidle_exit();
 		__cpu_suspend_exit();
 	}
diff --git a/arch/arm64/mm/kasan_init.c b/arch/arm64/mm/kasan_init.c
index b65a29440a0c..86ab780cff80 100644
--- a/arch/arm64/mm/kasan_init.c
+++ b/arch/arm64/mm/kasan_init.c
@@ -286,7 +286,11 @@ static void __init clear_next_level(int pgd_idx, int start, int end)
 	memset(&pudp[start], 0, (end - start) * sizeof(pud_t));
 }
 
+#ifdef CONFIG_AMLOGIC_VMAP
+void __init clear_shadow(u64 start, u64 end)
+#else
 static void __init clear_shadow(u64 start, u64 end)
+#endif
 {
 	int l = root_level_idx(start), m = root_level_idx(end);
 
diff --git a/fs/proc/meminfo.c b/fs/proc/meminfo.c
index d1627b530849..607d260d707a 100644
--- a/fs/proc/meminfo.c
+++ b/fs/proc/meminfo.c
@@ -23,6 +23,10 @@
 #include "internal.h"
 #include <trace/hooks/mm.h>
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
+
 void __attribute__((weak)) arch_report_meminfo(struct seq_file *m)
 {
 }
@@ -111,8 +115,13 @@ static int meminfo_proc_show(struct seq_file *m, void *v)
 	show_val_kb(m, "Slab:           ", sreclaimable + sunreclaim);
 	show_val_kb(m, "SReclaimable:   ", sreclaimable);
 	show_val_kb(m, "SUnreclaim:     ", sunreclaim);
+#ifdef CONFIG_AMLOGIC_VMAP
+	seq_printf(m, "KernelStack:    %8u kB\n",
+			get_vmap_stack_size() << (PAGE_SHIFT - 10));
+#else
 	seq_printf(m, "KernelStack:    %8lu kB\n",
 		   global_node_page_state(NR_KERNEL_STACK_KB));
+#endif
 #ifdef CONFIG_SHADOW_CALL_STACK
 	seq_printf(m, "ShadowCallStack:%8lu kB\n",
 		   global_node_page_state(NR_KERNEL_SCS_KB));
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 89afa3ae26b6..f6a5ea75f8ee 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1992,7 +1992,13 @@ extern unsigned long init_stack[THREAD_SIZE / sizeof(unsigned long)];
 #ifdef CONFIG_THREAD_INFO_IN_TASK
 # define task_thread_info(task)	(&(task)->thread_info)
 #elif !defined(__HAVE_THREAD_FUNCTIONS)
+#ifdef CONFIG_AMLOGIC_VMAP
+#define task_thread_info(task)		\
+	((struct thread_info *)(((unsigned long)(task)->stack) + \
+				  THREAD_INFO_OFFSET))
+#else
 # define task_thread_info(task)	((struct thread_info *)(task)->stack)
+#endif /* CONFIG_AMLOGIC_VMAP */
 #endif
 
 /*
diff --git a/include/linux/sched/task_stack.h b/include/linux/sched/task_stack.h
index 6c2fef89a4fd..31898c056273 100644
--- a/include/linux/sched/task_stack.h
+++ b/include/linux/sched/task_stack.h
@@ -27,11 +27,15 @@ static __always_inline void *task_stack_page(const struct task_struct *task)
 
 static __always_inline unsigned long *end_of_stack(const struct task_struct *task)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	return task->stack;
+#else /* CONFIG_AMLOGIC_VMAP */
 #ifdef CONFIG_STACK_GROWSUP
 	return (unsigned long *)((unsigned long)task->stack + THREAD_SIZE) - 1;
 #else
 	return task->stack;
 #endif
+#endif /* CONFIG_AMLOGIC_VMAP */
 }
 
 #elif !defined(__HAVE_THREAD_FUNCTIONS)
@@ -83,8 +87,12 @@ static inline void put_task_stack(struct task_struct *tsk) {}
 
 void exit_task_stack_account(struct task_struct *tsk);
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#define task_stack_end_corrupted(task) (false)
+#else
 #define task_stack_end_corrupted(task) \
 		(*(end_of_stack(task)) != STACK_END_MAGIC)
+#endif
 
 static inline int object_is_on_stack(const void *obj)
 {
diff --git a/include/linux/vmalloc.h b/include/linux/vmalloc.h
index 9660e0a2d82d..e3fcfda5a5e1 100644
--- a/include/linux/vmalloc.h
+++ b/include/linux/vmalloc.h
@@ -322,5 +322,11 @@ bool vmalloc_dump_obj(void *object);
 #else
 static inline bool vmalloc_dump_obj(void *object) { return false; }
 #endif
+#ifdef CONFIG_AMLOGIC_VMAP
+struct vm_struct *__get_vm_area_node(unsigned long size,
+		unsigned long align, unsigned long shift, unsigned long flags,
+		unsigned long start, unsigned long end, int node,
+		gfp_t gfp_mask, const void *caller);
+#endif
 
 #endif /* _LINUX_VMALLOC_H */
diff --git a/kernel/fork.c b/kernel/fork.c
index 128a70001c33..85be4597ffb3 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -115,6 +115,10 @@
 #include <asm/cacheflush.h>
 #include <asm/tlbflush.h>
 
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
+
 #include <trace/events/sched.h>
 
 #define CREATE_TRACE_POINTS
@@ -356,6 +360,7 @@ static void free_thread_stack(struct task_struct *tsk)
 
 #  else /* !CONFIG_VMAP_STACK */
 
+#ifndef CONFIG_AMLOGIC_VMAP
 static void thread_stack_free_rcu(struct rcu_head *rh)
 {
 	__free_pages(virt_to_page(rh), THREAD_SIZE_ORDER);
@@ -367,9 +372,18 @@ static void thread_stack_delayed_free(struct task_struct *tsk)
 
 	call_rcu(rh, thread_stack_free_rcu);
 }
+#endif
 
 static int alloc_thread_stack_node(struct task_struct *tsk, int node)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	void *stack = aml_stack_alloc(node, tsk);
+	if (stack) {
+		tsk->stack = stack;
+		return 0;
+	}
+	return -ENOMEM;
+#else /* CONFIG_AMLOGIC_VMAP */
 	struct page *page = alloc_pages_node(node, THREADINFO_GFP,
 					     THREAD_SIZE_ORDER);
 
@@ -378,11 +392,16 @@ static int alloc_thread_stack_node(struct task_struct *tsk, int node)
 		return 0;
 	}
 	return -ENOMEM;
+#endif /* CONFIG_AMLOGIC_VMAP */
 }
 
 static void free_thread_stack(struct task_struct *tsk)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	aml_stack_free(tsk);
+#else
 	thread_stack_delayed_free(tsk);
+#endif /* CONFIG_AMLOGIC_VMAP */
 	tsk->stack = NULL;
 }
 
@@ -524,6 +543,9 @@ void vm_area_free(struct vm_area_struct *vma)
 
 static void account_kernel_stack(struct task_struct *tsk, int account)
 {
+#ifdef CONFIG_AMLOGIC_VMAP
+	aml_account_task_stack(tsk, account);
+#else
 	if (IS_ENABLED(CONFIG_VMAP_STACK)) {
 		struct vm_struct *vm = task_stack_vm_area(tsk);
 		int i;
@@ -538,6 +560,7 @@ static void account_kernel_stack(struct task_struct *tsk, int account)
 		mod_lruvec_kmem_state(stack, NR_KERNEL_STACK_KB,
 				      account * (THREAD_SIZE / 1024));
 	}
+#endif
 }
 
 void exit_task_stack_account(struct task_struct *tsk)
@@ -1090,7 +1113,9 @@ void set_task_stack_end_magic(struct task_struct *tsk)
 	unsigned long *stackend;
 
 	stackend = end_of_stack(tsk);
+#ifndef CONFIG_AMLOGIC_VMAP
 	*stackend = STACK_END_MAGIC;	/* for overflow detection */
+#endif
 }
 
 static struct task_struct *dup_task_struct(struct task_struct *orig, int node)
diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index d9a740e1123e..8bc3ac5548c3 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -788,6 +788,7 @@ config SHRINKER_DEBUG
 config DEBUG_STACK_USAGE
 	bool "Stack utilization instrumentation"
 	depends on DEBUG_KERNEL
+	depends on !AMLOGIC_VMAP
 	help
 	  Enables the display of the minimum amount of free stack which each
 	  task has ever had available in the sysrq-T and sysrq-P debug output.
diff --git a/mm/kmemleak.c b/mm/kmemleak.c
index 0aecd537645a..2d0a76ef0d74 100644
--- a/mm/kmemleak.c
+++ b/mm/kmemleak.c
@@ -102,6 +102,9 @@
 #include <linux/kfence.h>
 #include <linux/kmemleak.h>
 #include <linux/memory_hotplug.h>
+#ifdef CONFIG_AMLOGIC_VMAP
+#include <linux/amlogic/vmap_stack.h>
+#endif
 
 /*
  * Kmemleak configuration and common defines.
@@ -1630,6 +1633,24 @@ static void kmemleak_cond_resched(struct kmemleak_object *object)
 	put_object(object);
 }
 
+#ifdef CONFIG_AMLOGIC_VMAP
+static void vmap_stack_scan(void *stack)
+{
+	int sum = 0;
+
+	if (likely(is_vmap_addr((unsigned long)stack))) {
+		while (!check_pte_exist((unsigned long)stack)) {
+			stack += PAGE_SIZE;
+			sum += PAGE_SIZE;
+			if (sum >= THREAD_SIZE)
+				break;
+		}
+	}
+	if (likely(sum < THREAD_SIZE))
+		scan_block(stack, stack + THREAD_SIZE - sum, NULL);
+}
+#endif
+
 /*
  * Scan data sections and all the referenced memory blocks allocated via the
  * kernel's standard allocators. This function must be called with the
@@ -1728,7 +1749,11 @@ static void kmemleak_scan(void)
 		for_each_process_thread(g, p) {
 			void *stack = try_get_task_stack(p);
 			if (stack) {
+			#ifdef CONFIG_AMLOGIC_VMAP
+				vmap_stack_scan(stack);
+			#else
 				scan_block(stack, stack + THREAD_SIZE, NULL);
+			#endif
 				put_task_stack(p);
 			}
 		}
-- 
2.50.1

